network/src/protocols/wire/handshake/v1/mod.rs

```diff
@@ -110,87 +135,105 @@ impl fmt::Display for MessagingProtocolVersion {
+//
+// HandshakeMsg
+//
+
+/// An enum to list the possible errors during the libra handshake negotiation
+#[derive(Debug, Error)]
+pub enum HandshakeError {
+    #[error("libra-handshake: the received message has an invalid chain id: {0}")]
+    InvalidChainId(ChainId),
+    #[error("libra-handshake: the received message has an invalid network id: {0}")]
+    InvalidNetworkId(NetworkId),
+    #[error("libra-handshake: could not find an intersection of supported protocol with the peer")]
+    NoCommonProtocols,
 }
 
-impl SupportedProtocols {
-    /// Returns a new SupportedProtocols struct that is an intersection.
-    fn intersection(self, other: SupportedProtocols) -> SupportedProtocols {
-        SupportedProtocols(self.0 & other.0)
-    }
+/// The HandshakeMsg contains a mapping from MessagingProtocolVersion suppported by the node to a
+/// bit-vector specifying application-level protocols supported over that version.
+#[derive(Clone, Deserialize, Serialize, Default)]
+pub struct HandshakeMsg {
+    pub supported_protocols: BTreeMap<MessagingProtocolVersion, SupportedProtocols>,
+    pub chain_id: ChainId,
+    pub network_id: NetworkId,
 }
 
 impl HandshakeMsg {
-    pub fn new(chain_id: ChainId, network_id: NetworkId) -> Self {
+    /// Useful function for tests
+    #[cfg(test)]
+    pub fn new_for_testing() -> Self {
+        let mut supported_protocols = BTreeMap::new();
+        supported_protocols.insert(
+            MessagingProtocolVersion::V1,
+            [ProtocolId::StateSynchronizerDirectSend].iter().into(),
+        );
         Self {
-            supported_protocols: Default::default(),
-            network_id,
-            chain_id,
+            chain_id: ChainId::test(),
+            network_id: NetworkId::Validator,
+            supported_protocols,
         }
     }
 
-    pub fn add(
-        &mut self,
-        messaging_protocol: MessagingProtocolVersion,
-        application_protocols: SupportedProtocols,
-    ) {
-        self.supported_protocols
-            .insert(messaging_protocol, application_protocols);
-    }
-
-    pub fn verify(&self, other: &HandshakeMsg) -> bool {
-        self.chain_id == other.chain_id && self.network_id == other.network_id
-    }
-
-    pub fn find_common_protocols(
+    /// This function:
+    /// 1. verifies that both HandshakeMsg are compatible and
+    /// 2. finds out the intersection of protocols that is supported
+    pub fn perform_handshake(
         &self,
         other: &HandshakeMsg,
-    ) -> Option<(MessagingProtocolVersion, SupportedProtocols)> {
-        // First, find the highest MessagingProtocolVersion supported by both nodes.
+    ) -> Result<(MessagingProtocolVersion, SupportedProtocols), HandshakeError> {
+        // verify that both peers are on the same chain
+        if self.chain_id != other.chain_id {
+            return Err(HandshakeError::InvalidChainId(other.chain_id));
+        }
+
+        // verify that both peers are on the same type of network
```

Comment:
I think these are self explanatory without the comments

Reply:
I like these as it reads like a specification, or can be matched with a specification more easily...